packages/uma-auth-client/src/hooks/useOAuth.tsx

```diff
@@ -0,0 +1,188 @@
+  persist(
+    (set, get) => ({
+      identityNpub:
+        "npub1scmpzl2ehnrtnhu289d9rfrwprau9z6ka0pmuhz6czj2ae5rpuhs2l4j9d", // TODO: remove this
+      identityRelayUrl: "wss://nos.lol", // TODO: remove this
+      redirectUri: "http://localhost:3001", // TODO: remove this
+      setToken: (token) =>
+        set({
+          token,
+        }),
+      initialOAuthRequest: async (uma) => {
+        const state = get();
+        const { codeVerifier, authUrl } = await getOAuthUrl(state, uma);
+        set({ codeVerifier });
+        window.location.href = authUrl;
+      },
+      oAuthTokenExchange: async (uma) => {
+        const state = get();
+        const result = await oAuthTokenExchange(state, uma);
+        set(result);
+      },
+    }),
+    {
+      name: "uma-connect",
+      partialize: (state) => ({
+        nwcConnectionUri: state.nwcConnectionUri,
+        codeVerifier: state.codeVerifier,
+      }),
+    },
+  ),
+);
+
+const getOAuthUrl = async (state: OAuthState, uma: string) => {
+  const { identityNpub, identityRelayUrl, redirectUri, codeVerifier } = state;
+  const discoveryDocument = await fetchDiscoveryDocument(uma);
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!,
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      base_url: new URL(discoveryDocument.authorization_endpoint).origin,
+    },
+  });
+
+  // TODO: get requiredCommands, optionalCommands, budget, expirationPeriod from client app
+  const requiredCommands = ["pay_invoice"];
+  const optionalCommands = ["get_balance", "list_transactions"];
+  const budget = "10.USD%2Fmonthly";
+  const expirationPeriod = "year";
+
+  const authUrl = await authClient.generateAuthURL({
+    authorizeUrl: `${discoveryDocument.authorization_endpoint}?required_commands=${requiredCommands.join(",")}&optional_commands=${optionalCommands.join(",")}&budget=${budget}&expiration_period=${expirationPeriod}`,
+    code_challenge_method: "S256",
+  });
+
+  return {
+    codeVerifier: authClient.code_verifier || "",
+    authUrl,
+  };
+};
+
+const oAuthTokenExchange = async (state: OAuthState, uma: string) => {
+  const {
+    identityNpub,
+    identityRelayUrl,
+    redirectUri,
+    codeVerifier,
+    token,
+    nwcConnectionUri,
+    nwcExpiresAt,
+  } = state;
+  const discoveryDocument = await fetchDiscoveryDocument(uma);
+
+  // If we have a connection URI and a token that hasn't expired, we don't need to do anything
+  if (
+    nwcConnectionUri &&
+    token &&
+    (!token.expiresAt || Date.now() < token.expiresAt)
+  ) {
+    return state;
+  }
+
+  const authClient = new auth.OAuth2User({
+    client_id: `${identityNpub} ${identityRelayUrl}`,
+    callback: redirectUri!, // TODO: get this from client app
+    scopes: [],
+    user_agent: "uma-connect",
+    request_options: {
+      endpoint: discoveryDocument.token_endpoint,
```

Comment:
ugh... it looks like the alby sdk wipes this out with their own setting :-/. For now in this pr, let's use the base_url since it sets the endpoint to /oauth/token. We may need to figure out a way to work around this in parallel though...